File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    if number == N:
        return 1
    values = {}
    tmp = 0
    for i in range(1, 9):
        tmp += N
        if number == tmp: 
            return i
        values[i] = {tmp}
        tmp *= 10
    
    eval = 0
    for i in range(2, 9):
        for j in range(1, i // 2 + 1):
            for x in values[i - j]:
                for y in values[j]:
                    for z in range(6):
                        if z == 0:
                            eval = x + y
                        elif z == 1:
                            eval = x * y
                        elif z == 2:
                            eval = x - y
                        elif z == 3:
                            eval = y - x
                        elif z == 4:
                            eval = x // y
                        elif z == 5:
                            eval = y // x
                        if eval > 0:
                            if eval == number:
                                return i
                            values[i].add(eval)
    return -1

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        if (N == number) return 1;
        List<Set<Integer>> dict = new ArrayList<>();
        int tmp = 0;
        for (int i = 0; i <= 8; i++) {
            dict.add(new HashSet<>());
            if (i >= 1) {
                tmp += N;
                if (tmp == number) return i;
                dict.get(i).add(tmp);
                tmp *= 10;
            }
        }
        int eval = 0;
        for (int i = 2; i <= 8; i++) {
            for (int j = 1; j < i/2 + 1; j++) {
                for (int val1 : dict.get(i - j)) {
                    for (int val2 : dict.get(j)) {
                        for (int k = 0; k < 6; k++) {
                            if (k == 0) {
                                eval = val1 + val2;
                            }
                            else if (k == 1) {
                                eval = val1 - val2;
                            }
                            else if (k == 2) {
                                eval = val2 - val1;
                            }
                            else if (k == 3) {
                                eval = val1 / val2;
                            }
                            else if (k == 4) {
                                eval = val2 / val1;
                            }
                            else if (k == 5) {
                                eval = val1 * val2;
                            }
                            if (eval > 0) {
                                if (eval == number) return i;
                                dict.get(i).add(eval);
                            }
                        }
                    }
                }
            }
        }
        
        return -1;
    }
}